v0/destinations/user/utils.js

```diff
@@ -0,0 +1,643 @@
+const get = require("get-value");
+const {
+  CustomError,
+  toUnixTimestamp,
+  constructPayload,
+  getIntegrationsObj,
+  getValueFromMessage,
+  getDestinationExternalID,
+  getFieldValueFromMessage,
+  isDefinedAndNotNullAndNotEmpty
+} = require("../../util");
+const { httpGET, httpPOST, httpPUT } = require("../../../adapters/network");
+const {
+  processAxiosResponse
+} = require("../../../adapters/utils/networkUtils");
+const {
+  BASE_ENDPOINT,
+  MAPPING_CONFIG,
+  CONFIG_CATEGORIES,
+  groupSourceKeys,
+  identifySourceKeys
+} = require("./config");
+
+/**
+ * Returns updated User.com url
+ * @param {*} endpoint
+ * @param {*} appSubdomain
+ * @returns
+ */
+const prepareUrl = (endpoint, appSubdomain) => {
+  return endpoint.replace("appSubdomainName", appSubdomain);
+};
+
+/**
+ * Returns company address in string format
+ * @param {*} endpoint
+ * @param {*} appSubdomain
+ * @returns
+ */
+const prepareCompanyAddress = address => {
+  let companyAddress = "";
+  const keys = Object.keys(address);
+  keys.forEach(key => {
+    companyAddress += address[key];
+  });
+  return companyAddress;
+};
+
+/**
+ * Returns the remaining keys from traits
+ * Remaining keys : keys which is not included in webapp configuration mapping and not included in source-dest keys file
+ * @param {*} traits
+ * @param {*} sourceKeys
+ * @returns
+ */
+const getRemainingTraits = (traits, sourceKeys) => {
+  const properties = {};
+  const keys = Object.keys(traits);
+  keys.forEach(key => {
+    if (!sourceKeys.includes(key)) {
+      properties[key] = traits[key];
+    }
+  });
+  return properties;
+};
+
+/**
+ * Returns the user attributes(configured from webapp mapping) and remaining attributes(which is not configured anywhere)
+ * @param {*} userAttributesMap
+ * @param {*} message
+ * @returns
+ */
+const getCustomUserAttributes = (userAttributesMap, message) => {
+  const properties = {};
+  let traits = getFieldValueFromMessage(message, "traits");
+  const contextTraits = get(message, "context.traits");
+  traits = { ...traits, ...contextTraits };
+  const sourceKeys = identifySourceKeys;
+
+  userAttributesMap.forEach(attribute => {
+    const { from, to } = attribute;
+    if (traits[from]) {
+      properties[to] = traits[from];
+      sourceKeys.push(from);
+    }
```

Comment:
the functions for fetching the user, event and company attributes are almost same. Can we brainstorm it a little bit and merge these three if possible.

Reply:
also we have an util function called `getHashFromArray`, lets explore if the input pattern that we are having, that function matches it. If yes, we can use that as well.